src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -0,0 +1,276 @@
+            bool hasRecursive = includeGlobPathAbsolute.Contains("**", StringComparison.Ordinal);
+
+            // no globbing
+            if (firstGlob < 0)
+            {
+                return new Enumeration() {
+                    RootPath = Path.GetDirectoryName(includeGlobPathAbsolute),
+                    Pattern = Path.GetFileName(includeGlobPathAbsolute),
+                    Depth = SearchOption.TopDirectoryOnly
+                };
+            }
+            else
+            {
+                int rootDirLength = includeGlobPathAbsolute.Substring(0,firstGlob).LastIndexOfAny( new [] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar});
+                return new Enumeration() {
+                    RootPath = includeGlobPathAbsolute.Substring(0,rootDirLength),
+                    Pattern = "*",
+                    Depth = hasRecursive ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly
+                };
+            }
+        }
+
+        internal static void CheckKeySegment(string keySegment)
+        {
+            if (keySegment.Equals("*", StringComparison.Ordinal))
+            {
+                throw new ArgumentException("`*` is a reserved key segment. For path glob, use `./*`.");
+            }
+            else if (keySegment.Equals(Fingerprint.Wildcard, StringComparison.Ordinal))
+            {
+                throw new ArgumentException("`**` is a reserved key segment. For path glob, use `./**`.");
+            }
+            else if (keySegment.First() == '\'')
+            {
+                throw new ArgumentException("A key segment cannot start with a single-quote character`.");
+            }
+            else if (keySegment.First() == '`')
+            {
+                throw new ArgumentException("A key segment cannot start with a backtick character`.");
+            }
+        }
+
+        public static Fingerprint EvaluateKeyToFingerprint(
+            AgentTaskPluginExecutionContext context,
+            string filePathRoot,
+            IEnumerable<string> keySegments)
+        {
+            var sha256 = new SHA256Managed();
+
+            string defaultWorkingDirectory = context.Variables.GetValueOrDefault(
+                "system.defaultworkingdirectory" // Constants.Variables.System.DefaultWorkingDirectory
+                )?.Value;
+
+            var resolvedSegments = new List<string>();
+
+            foreach (string keySegment in keySegments)
+            {
+                CheckKeySegment(keySegment);
+            }
+
+            foreach (string keySegment in keySegments)
+            {
+                if (IsPathy(keySegment))
+                {
+                    context.Verbose($"Interpretting `{keySegment}` as a path.");
+
+                    string[] pathRules = keySegment.Split(new []{','}, StringSplitOptions.RemoveEmptyEntries).Select(s => s.Trim()).ToArray();
+                    string[] includeRules = pathRules.Where(p => !p.StartsWith('!')).ToArray();
```

Comment:
>'!' [](start = 79, length = 3)

What does '!' mean in terms of a path segment?

Reply:
"except" - same as gitignore and artifactignore https://docs.microsoft.com/en-us/azure/devops/artifacts/reference/artifactignore?view=azure-devops

Reply:
Does Minimatch already handle the commas?

---
In reply to: [306966358](https://github.com/microsoft/azure-pipelines-agent/pull/2358#discussion_r306966358) [](ancestors = 306966358)